contracts/src/contracts/ValidatorManager/BalancerValidatorManager.sol

```diff
@@ -0,0 +1,320 @@
+        /// @notice Validators pending weight updates
+        mapping(bytes32 validationID => bytes32 messageID) validatorPendingWeightUpdate;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("suzaku.storage.BalancerValidatorManager")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 public constant BALANCER_VALIDATOR_MANAGER_STORAGE_LOCATION =
+        0x9d2d7650aa35ca910e5b713f6b3de6524a06fbcb31ffc9811340c6f331a23400;
+
+    // solhint-disable func-name-mixedcase, ordering
+    function _getBalancerValidatorManagerStorage()
+        private
+        pure
+        returns (BalancerValidatorManagerStorage storage $)
+    {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            $.slot := BALANCER_VALIDATOR_MANAGER_STORAGE_LOCATION
+        }
+    }
+
+    modifier onlySecurityModule() {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (!$.securityModules.contains(msg.sender)) {
+            revert BalancerValidatorManager__UnknownSecurityModule(msg.sender);
+        }
+        _;
+    }
+
+    constructor(
+        ICMInitializable init
+    ) {
+        if (init == ICMInitializable.Disallowed) {
+            _disableInitializers();
+        }
+    }
+
+    function initialize(
+        ValidatorManagerSettings calldata settings,
+        address initialOwner,
+        address initialSecurityModule,
+        uint64 initialSecurityModuleWeight
+    ) external initializer {
+        __BalancerValidatorManager_init(
+            settings, initialOwner, initialSecurityModule, initialSecurityModuleWeight
+        );
+    }
+
+    function __BalancerValidatorManager_init(
+        ValidatorManagerSettings calldata settings,
+        address initialOwner,
+        address initialSecurityModule,
+        uint64 initialSecurityModuleWeight
+    ) internal onlyInitializing {
+        __ValidatorManager_init(settings);
+        __Ownable_init(initialOwner);
+        _setupSecurityModule(initialSecurityModule, initialSecurityModuleWeight);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function __BalancerValidatorManager_init_unchained() internal onlyInitializing {}
+
+    // solhint-enable func-name-mixedcase
+
+    /// @inheritdoc IBalancerValidatorManager
+    function setupSecurityModule(address securityModule, uint64 maxWeight) external onlyOwner {
+        _setupSecurityModule(securityModule, maxWeight);
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function initializeValidatorRegistration(
+        ValidatorRegistrationInput calldata registrationInput,
+        uint64 weight
+    ) external onlySecurityModule returns (bytes32 validationID) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (!$.securityModules.contains(msg.sender)) {
+            revert BalancerValidatorManager__SecurityModuleNotRegistered(msg.sender);
```

Comment:
Shouldn't this be `BalancerValidatorManager__NewWeightIsZero`? Or eventually a new `BalancerValidatorManager__MaxWeightIsZero`

Reply:
I don't quite get that. This condition checks that only registered SecurityModules can register new validators.